### PR TITLE
Make ActorHandles pickleable, also make proper ActorHandle and ActorC…

### PR DESCRIPTION
**Pull request by robertnishihara**
…lass classes.

This is intended to cleanup some of the Python actor code.

Note that should enable named actors via https://github.com/ray-project/ray/issues/1424#issuecomment-386764958.

### CODE REVIEW

File: python/ray/actor.py [hunk 1]
```diff
@@ -854,12 +831,7 @@ def make_actor(cls, resources, checkpoint_interval, actor_method_cpus):
     # Modify the class to have an additional method that will be used for
     # terminating the worker.
     class Class(cls):
-        def __ray_terminate__(self, actor_id):
-            # Record that this actor has been removed so that if this node
-            # dies later, the actor won't be recreated. Alternatively, we could
-            # remove the actor key from Redis here.
-            ray.worker.global_worker.redis_client.hset(b"Actor:" + actor_id,
```
Comment (robertnishihara) [inline python/ray/actor.py]: Is this key in Redis used anywhere?

File: python/ray/actor.py [hunk 2]
```diff
@@ -494,357 +494,433 @@ def remote(self, *args, **kwargs):
...
+            "method_signatures":
+            self._ray_method_signatures,
+            "actor_creation_dummy_object_id":
+            self._ray_actor_creation_dummy_object_id.id(),
+            "actor_method_cpus":
+            self._ray_actor_method_cpus,
+        }
+
+        if ray_forking:
+            state["forked_actor_handle_id"] = compute_actor_handle_id(
```
Comment (robertnishihara) [inline python/ray/actor.py]: I'm using this method to compute actor handle IDs for "forked" actor handles. However, if a user pickles and unpickles the actor handle, then I'm generating the actor handle ID randomly.

**Comment by robertnishihara** [comment]
Maybe the right way to do actor handle ID generation (for pickled handles) is the following. Suppose we have

```python
actor_handle = Actor.remote()

# Defining this remote function, which closes over actor_handle, will
# cause actor_handle to be pickled. The remote function will be broadcast
# to all the workers and unpickled there.
@ray.remote
def f():
    actor_handle.method.remote()

f.remote()
```

Instead of generating a random actor handle ID on each worker when the remote function is broadcast, maybe we should instead leave the actor handle in an "uninitialized" state, and then whenever a `f.remote()` task executes, we generate a new actor handle ID deterministically from the actor ID and the current task ID. In other words, we want it to be roughly equivalent to calling `f.remote(actor_handle)`. Doing this will cause a proliferation of actor handles, so we'll need to make sure that is handled efficiently.

cc @stephanie-wang

**Comment by robertnishihara** [comment]
@stephanie-wang one use case is that you want to use the actor handle on a **different driver** (e.g., "named actors".

Another use case, I think, in RLlib/Tune is that certain user-defined functions have to have pre-specified signatures and the user isn't really able to just pass in an actor handle because that would require changing the signature and piping it through a bunch of RLlib abstractions, but they can define a function that just closes over the actor handle.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**